### ISSUE TITLE
TypeError: Cannot read properties of undefined (reading 'category') on current version

### ISSUE THREAD

**Issue by roguh** [microsoft/pyright#2892]
**Describe the bug**
I am running `pyright` on all my files on a work repository and it is crashing with `An internal error occurred while type checking file "/home/hugo/src/powerflex_edge_traffic_manager/src/traffic_manager/types.py": TypeError: Cannot read properties of undefined (reading 'category')`

Full traceback:
```
$ pipenv run pyright src
No configuration file found.
pyproject.toml file found at /home/hugo/src/powerflex_edge_traffic_manager.
Loading pyproject.toml file at /home/hugo/src/powerflex_edge_traffic_manager/pyproject.toml
Pyproject file "/home/hugo/src/powerflex_edge_traffic_manager/pyproject.toml" is missing "[tool.pyright]" section.
stubPath /home/hugo/src/powerflex_edge_traffic_manager/typings is not a valid directory.
Assuming Python platform Linux
Searching for source files
File or directory "/home/hugo/src/powerflex_edge_traffic_manager/--" does not exist.
Found 28 source files
An internal error occurred while type checking file "/home/hugo/src/powerflex_edge_traffic_manager/src/traffic_manager/types.py": TypeError: Cannot read properties of undefined (reading 'category')
    at isFunction (/home/hugo/.npm/_npx/34d0616bec88325f/node_modules/pyright/dist/pyright-internal/src/analyzer/types.ts:2090:17)
    at Object.canOverrideMethod (/home/hugo/.npm/_npx/34d0616bec88325f/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:22081:14)
    at callback (/home/hugo/.npm/_npx/34d0616bec88325f/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluatorWithTracker.ts:136:58)
    at s.timeOperation (/home/hugo/.npm/_npx/34d0616bec88325f/node_modules/pyright/dist/pyright-internal/src/common/timing.ts:44:28)
    at run (/home/hugo/.npm/_npx/34d0616bec88325f/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluatorWithTracker.ts:43:46)
    at Object.canOverrideMethod (/home/hugo/.npm/_npx/34d0616bec88325f/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluatorWithTracker.ts:136:13)
    at forEach (/home/hugo/.npm/_npx/34d0616bec88325f/node_modules/pyright/dist/pyright-internal/src/analyzer/checker.ts:4117:50)
    at Map.forEach (<anonymous>)
    at A._validateBaseClassOverrides (/home/hugo/.npm/_npx/34d0616bec88325f/node_modules/pyright/dist/pyright-internal/src/analyzer/checker.ts:4041:34)
    at A.visitClass (/home/hugo/.npm/_npx/34d0616bec88325f/node_modules/pyright/dist/pyright-internal/src/analyzer/checker.ts:322:22)
Error performing analysis: TypeError: Cannot read properties of undefined (reading 'category')
    at isFunction (/home/hugo/.npm/_npx/34d0616bec88325f/node_modules/pyright/dist/pyright-internal/src/analyzer/types.ts:2090:17)
    at Object.canOverrideMethod (/home/hugo/.npm/_npx/34d0616bec88325f/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:22081:14)
    at callback (/home/hugo/.npm/_npx/34d0616bec88325f/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluatorWithTracker.ts:136:58)
    at s.timeOperation (/home/hugo/.npm/_npx/34d0616bec88325f/node_modules/pyright/dist/pyright-internal/src/common/timing.ts:44:28)
    at run (/home/hugo/.npm/_npx/34d0616bec88325f/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluatorWithTracker.ts:43:46)
    at Object.canOverrideMethod (/home/hugo/.npm/_npx/34d0616bec88325f/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluatorWithTracker.ts:136:13)
    at forEach (/home/hugo/.npm/_npx/34d0616bec88325f/node_modules/pyright/dist/pyright-internal/src/analyzer/checker.ts:4117:50)
    at Map.forEach (<anonymous>)
    at A._validateBaseClassOverrides (/home/hugo/.npm/_npx/34d0616bec88325f/node_modules/pyright/dist/pyright-internal/src/analyzer/checker.ts:4041:34)
    at A.visitClass (/home/hugo/.npm/_npx/34d0616bec88325f/node_modules/pyright/dist/pyright-internal/src/analyzer/checker.ts:322:22)
```

**To Reproduce**
Run pyright on specific code.

**Expected behavior**
I expect Pyright not to crash when run on code that it has checked successfully before.

**Screenshots or Code**
The traceback mentions my types.py file, but not a specific line number. I cannot post the full file because it is proprietary code. I can work with you to find out what line is causing the problem.

**VS Code extension or command-line**
From the CLI in GitHub actions and locally.

```
$ pipenv run pyright --version
pyright 1.1.211
$ pipenv run python -c 'import pyright;print(pyright.__version__)'
0.0.12
```

**Additional context**
I saw https://github.com/microsoft/pyright/issues/872 and this did not fix my issue as I am using a new version of pyright.

**Comment by erictraut** [comment]
Thanks for reporting the problem. I think I'll need your help tracking this down.

A few questions for you:
1. Do you know if this crash started occurring with a recent release of pyright? Was it able to type check this code base previously without a crash?
2. The stack crawl seems to indicate that the problem involves a class that's declared in the file `types.py` that derives from another class and overrides one of the methods in the base class. Furthermore, the method declaration in the base class is overloaded (i.e. has one or more `@overload` signatures on it). Does that help narrow down the issue?

Ah, I think I may have found it. Do you use `TypedDict` in your code? If so, do you have a `TypedDict` with zero fields defined? And then do you derive another `TypedDict` from that empty `TypedDict`, like this?

```python
from typing_extensions import TypedDict
class TD1(TypedDict): pass
class TD2(TD1):
    x: int
```

This repros the crash you're seeing. The fix for this is simple, but I want to make sure that it's the same case as the one you're hitting in your code.

**Comment by roguh** [comment]
I have that exact code!

```
from typing import TypedDict

class RedactedBase(TypedDict):
    pass


class RedactedClass(RedactedBase):
    redacted: bytes
```

We use NewTypes and other types in this, I can send more details if necessary.

**Comment by roguh** [comment]
What is the fix?

**Comment by erictraut** [comment]
Great! Then I can confirm that the next release of pyright will address this issue. 

If you're interested in the details... Pyright tracks overloads using a data structure that includes an array of function signatures (one for each overload). It assumes that this array will contain at least one element. The code that handles `TypedDict` needs to synthesize certain methods including `pop` and `setdefault`, and it generates one overloaded signature for each field in your `TypedDict`. If you have zero fields in your `TypedDict`, the code was incorrectly creating an overloaded function with zero overloads. This later caused a crash in a piece of code that assumed there would be at least one. Here's the [commit](https://github.com/microsoft/pyright/commit/9134e2a76c8f0add06a3e6481b6d2be947a4ad2f) if you're interested.

**Comment by roguh** [comment]
Thank you for the prompt responses! This is very informative as well. We'll wait for the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.212, which I just published. It will also be included in the next release of pylance.

**Comment by roguh** [comment]
Thank you!
